src/datastore/postgres-store.ts

```diff
@@ -6816,7 +6905,20 @@ export class PgDataStore
           tx_index, tx_id, status, canonical,
           index_block_hash, parent_index_block_hash, microblock_hash, microblock_sequence, microblock_canonical
         ) values($1, $2, $3, $4, $5, $6, $7, $8,$9, $10, $11, $12, $13, $14, $15)
-        `,
+        ON CONFLICT ON CONSTRAINT unique_name_tx_id_index_block_hash_microblock_hash DO
```

Comment:
Could you provide some context around the condition here? Much of the BNS code is just simple data transforms and I'm having a hard time following the changes to the actual logic, but it seems like these unique constraints are the main ones?

So in this case, what are the conditions around a new name event containing a duplicate `(txid, index_block_hash, microblock_hash)`?

Reply:
It would be super helpful to have some tests that replicate the behavior that this is solving

Reply:
In practice, there are two main cases in which we get duplicate names/subdomains/zonefiles:

1. **When we get a `name-register` followed by a `name-update` BNS event**
    The registration sends an attachment with an empty zonefile, and the update sends an attachment with the real zonefile usually a few blocks after. In the old code, each of these events inserted duplicate values because the attachments referenced the same tx_id/index_block_hash/microblock_hash instead of just updating the temporary registration rows (as it should be). This translates to literally thousands of duplicate and/or garbage rows in all three tables. When the old code DID update, though, it incorrectly updated based on the zonefile_hash which will *obviously* change when the name-update event gets sent (the zonefile changed, so the hash changed).
2. **When we get duplicate attachments**
    In the last replays we've done, I've seen attachment messages that contain duplicate entries for name updates (possibly from re-orgs or other node network issues). Today we process those as-is, inserting duplicates of everything.

> It would be super helpful to have some tests that replicate the behavior that this is solving

I'll definitely add some tests for this